planning/MartenV4.md

```diff
@@ -0,0 +1,465 @@
+
+My thought is to designate one (or maybe a few?) .Net type as the "metadata type" for your application like maybe this one:
+
+```
+    public class MyMetadata
+    {
+        public Guid CorrelationId { get; set; }
+        public string UserId { get; set; }
+    }
+```
+Maybe that gets added to the `StoreOptions` something like:
+
+```
+var store = DocumentStore.For(x => {
+    // other stuff
+
+    // This would direct Marten to add extra columns to
+    // the documents and events for the metadata properties
+    // on the MyMetadata type.
+
+    // This would probably be a fluent interface to optionally fine tune
+    // the storage and applicability -- i.e., to all documents, to events, etc.
+    x.UseMetadataType<MyMetadata>();
+});
+```
+
+Then at runtime, you'd do something like:
+
+```
+session.UseMetadata<MyMetadta>(metadata);
+```
+
+Either through docs or through the new, official .Net Core integration, we have patterns to have that automatically set upon
+new `DocumentSession` objects being created from the IoC to make the tracking be seemless.
+
+## Extract Generic Database Helpers to its own Library
+
+* Pull everything to do with Schema object generation, difference detection, and DDL generation to a separate library (`IFeatureSchema`, `ISchemaObject`, etc.). Mostly to clean out the main library, but also because this code could easily be reused outside of Marten. Separating it out might make it easier to test and extend that functionality, which is something that occasionally gets requested. There's also the possibility of further breaking that into abstractions and implementations for the long run of getting us ready for Sql Server or other database engine support. **The tests for this functionality are slow, and rarely change. It would be advantageous to get this out of the main Marten library and testing project**.
+
+* Pull the ADO.Net helper code like `CommandBuilder` and the extension methods into a small helper library somewhere else (I'm nominating the [Baseline](https://jasperfx.github.io/baseline) repository). This code is copied around to other projects as it is, and it's another way of getting stuff out of the main library and the test suite.
+
+Track this work in [this GitHub issue](https://github.com/JasperFx/marten/issues/1467).
+
+## F# Improvements
+
+We'll have a virtual F# subcommittee to be watching this work for F#-friendliness:
+
+* [Fred](https://github.com/wastaz)
+* [Chet Husk](https://twitter.com/ChetHusk)
+* [Jimmy Byrd](https://github.com/TheAngryByrd)
+
+## HostBuilder Integration
+
+We'll bring Joona-Pekka Kokko's [ASP.Net Core integration library](https://github.com/jokokko/marten.aspnetcore) into the main repository and make that the officially blessed and documented recipe for integrating Marten into .Net Core applications based on the `HostBuilder` in .Net Core. I suppose we could also multi-target `IWebHostBuilder` for ASP.Net Core 2.*.
```

Comment:
I think that we should be good with supporting only 3.1 and .NET 5. 3.1 is LTS, all below (also 2.2) are not officially supported.